### ISSUE TITLE
Revisiting Procedural City: Flipping Normal and Inset issue

### ISSUE THREAD

**Issue by enzyme69** [nortikin/sverchok#902]
![screen shot 2016-10-08 at 11 17 02](https://cloud.githubusercontent.com/assets/3952687/19208664/ddde24e8-8d48-11e6-9ecb-044f54c86ba8.png)

I am revisiting the idea of "voronoi city" and want to make Greeble suing Sverchok. I am reading my old articles on this. And might create video as well.

Google Search reference:
- blender displacement detailed greeble
- hongkong kowloon hidden city
- Maya QTown --> someone really should recreate something like this using Sverchok.

Ok, basically the issue I still get is the Voronoi output normal that I cannot control. Some of the faces are flipped. I have to read old thread in this github....

We have Inset Special node, is this node really working? Looks like it is working sometimes.

So far, Extrude Face Separate node is the most promising. It does not do multiple extrude per face but probably I could use Adaptive.

I am sure this can be done nicely.

**Comment by enzyme69** [comment]
See also:
https://github.com/nortikin/sverchok/issues/185

**Comment by enzyme69** [comment]
![screen shot 2016-10-08 at 11 40 49](https://cloud.githubusercontent.com/assets/3952687/19208893/1cc62d56-8d4c-11e6-9829-d870c9968797.png)
I remember I was using Adaptive Poly and also has this face flipped issue.

Maybe I should separate the base face.

**Comment by zeffii** [comment]
_Inset Special_ only works on quads and tris.

btw, use Export Gist to share these light weight examples.
![really](https://cloud.githubusercontent.com/assets/619340/19211423/13464578-8d3c-11e6-9631-2110bf612682.gif)

**Comment by zeffii** [comment]
I would probably add a `recalc normals` after the fill holes, if possible. Tho normals are a little trickier to determine when a polygon has no direct neighbours (as might be the case when you have many disjoint faces in a mesh), the direction of the face normal will be determined by the order of the indices in the face.

**Comment by zeffii** [comment]
https://gist.github.com/79c584b3c1ae0d9cb7e88e1931b7359f

**Comment by zeffii** [comment]
all you have to do to load that is copy the url.

![likeso](https://cloud.githubusercontent.com/assets/619340/19211636/87a75a9c-8d41-11e6-8189-af4bdf103806.gif)

**Comment by zeffii** [comment]
it seems the adaptive polygons node may be where the fix needs to be here.. those polygons extruded are not randomly oriented.. (you could test the normals to prove that)

![image](https://cloud.githubusercontent.com/assets/619340/19211882/2e971d92-8d47-11e6-9c0e-b12b04ae62a6.png)

https://gist.github.com/d25d7ba0dcaa17aa54d61f58ab9e6f87

**Comment by zeffii** [comment]
![image](https://cloud.githubusercontent.com/assets/619340/19212087/5baf9b4e-8d4a-11e6-8121-7b0247b9010d.png)

https://gist.github.com/3c467ba3be191f56d24baf467096047d

**Comment by zeffii** [comment]
now you'll  notice that the normals only start to getting interpretted as flipped by the Adaptive Polygons node when the polygon becomes increasingly distorted. Let's look at the node's code to see how it determines the direction to work on.

The problem appears to be the recalc_normals bmesh operator that adaptive polygons also uses..

https://gist.github.com/071ad6e5f1ae316f68050ad316c6e1e0

![image](https://cloud.githubusercontent.com/assets/619340/19213109/ba31a8b6-8d62-11e6-8ed9-0ba3f937b523.png)

https://gist.github.com/9c5a508f1ba0f8f5518e0f70f64764a1

![image](https://cloud.githubusercontent.com/assets/619340/19213142/935795f6-8d63-11e6-95b0-8cccef1bc550.png)

**Comment by zeffii** [comment]
and then.. the IndeViewer to deal the final blow...

![image](https://cloud.githubusercontent.com/assets/619340/19213160/f9e6416e-8d63-11e6-9dce-ad67e77d8b56.png)

**Comment by zeffii** [comment]
I think the solution is that we write an algorithm for recalc normals that takes into account neighbouring (but possibly disjoint) faces.. there could be an initial brain dead algorithm that can assume all faces are intended to be oriented in one direction, and flips the index order of all faces that deviated from the first found polygon (it should also have a flip switch as an argument, so you don't need an additional flip-normals node behind the new algorithm/node )

**Comment by zeffii** [comment]
![image](https://cloud.githubusercontent.com/assets/619340/19213247/aaf0c662-8d66-11e6-9915-0579e29f0d41.png)

**Comment by nortikin** [comment]
extrude separate faces works better here, i guess

**Comment by zeffii** [comment]
maybe best to start a new Issue about a potential alternative recalc_normals algorithm..

**Comment by zeffii** [comment]
from your video @enzyme69 

https://gist.github.com/9585ab850665c20524e9a21b6c30636c

**Comment by enzyme69** [comment]
the only simple solution I could think, is to make all faces as if they are
separate and flip or control the face normal using vector or list of vector.

but this is a nice read I will put it as pending.


-- 


------------------------------------------------------------
- Jimmy Gunawan
http://blendersushi.blogspot.com.au
http://mayaspiral.blogspot.com.au
http://houdoodles.blogspot.com.au
http://puppetar.blogspot.com.au

**Comment by zeffii** [comment]
I think the reason for flipping was answered, and a solution is the works but will take a while..
